File: apps/ios/BitOS/Features/Profile/ProfileEditSheet.swift
import BusinessCore
import PhotosUI
import SwiftUI

/// Profile editing sheet: bounded fields → signed kind-0 through the
/// receipt machine.
///
/// UX parity with the legacy Flutter editor (`settings/pages/profile_page`):
/// a live header preview (banner + "Change banner" pill, hex avatar with a
/// camera chip → pick → center-crop → upload → URL) above the full kind-0
/// field form.
struct ProfileEditSheet: View {
    let publisher: NotePublisher
    let initialProfile: ProfileMetadata?
    let onClose: () -> Void
    @Environment(IdentityStore.self) private var identity
    @State private var name = ""
    @State private var displayName = ""
    @State private var about = ""
    @State private var nip05 = ""
    @State private var lud16 = ""
    @State private var picture = ""
    @State private var banner = ""
    @State private var website = ""
    @State private var published = false
    @State private var loaded = false
    @State private var uploadingTarget: String?
    @State private var uploadError: String?
    @State private var avatarItem: PhotosPickerItem?
    @State private var bannerItem: PhotosPickerItem?
    @State private var showAvatarPicker = false
    @State private var showBannerPicker = false
    /// Legacy step flow: pick target → source sheet (camera/library).
    @State private var sourceSheetTarget: SourceTarget?
    @State private var showCamera = false
    @State private var cameraTarget: String?
    @Environment(AppEnvironment.self) private var environment

    /// Legacy ImageCropEditor default-crop parity: normalize EXIF
    /// orientation, center-crop to the target aspect, downscale, re-encode
    /// JPEG, Blossom upload, fill the URL field.
    private func uploadPicked(_ target: String, _ data: Data?) async {
        guard let data, let raw = UIImage(data: data) else { return }
        let ui = raw.normalizedForCrop()
        let isAvatar = target == "avatar"
        let targetW: CGFloat = isAvatar ? 512 : 1500
        let targetH: CGFloat = isAvatar ? 512 : 500
        let targetAspect = targetW / targetH
        let srcW = ui.size.width
        let srcH = ui.size.height
        var cropRect: CGRect
        if srcW / srcH > targetAspect {
            let h = srcH
            let w = srcH * targetAspect
            cropRect = CGRect(x: (srcW - w) / 2, y: 0, width: w, height: h)
        } else {
            let w = srcW
            let h = srcW / targetAspect
            cropRect = CGRect(x: 0, y: (srcH - h) / 2, width: w, height: h)
        }
        guard let cg = ui.cgImage?.cropping(to: cropRect) else { return }
        let cropped = UIImage(cgImage: cg)
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        let rendered = UIGraphicsImageRenderer(size: CGSize(width: targetW, height: targetH), format: format).image { _ in
            cropped.draw(in: CGRect(x: 0, y: 0, width: targetW, height: targetH))
        }
        guard let jpeg = rendered.jpegData(compressionQuality: 0.88) else { return }
        uploadingTarget = target
        uploadError = nil
        do {
            let receipt = try await BlossomUploader().upload(
                bytes: jpeg, mimeType: "image/jpeg",
                identity: environment.identityStore,
                serverUrl: "https://blossom.primal.net"
            )
            if isAvatar { picture = receipt.url } else { banner = receipt.url }
        } catch {
            uploadError = "Upload failed — check your connection and try again."
        }
        uploadingTarget = nil
    }

    var body: some View {
        VStack(spacing: 0) {
            pageHeader
            content
                .padding(.horizontal, BitOSTheme.Spacing.screen)
                .padding(.bottom, BitOSTheme.Spacing.base)
        }
        .frame(maxHeight: .infinity, alignment: .top)
        .background(BitOSTheme.background)
        .preferredColorScheme(.dark)
        .sheet(item: $sourceSheetTarget) { target in
            sourceSheet(target: target.id)
        }
        .fullScreenCover(isPresented: $showCamera) { cameraPicker }
        .onAppear {
            guard !loaded else { return }
            loaded = true
            name = initialProfile?.name ?? ""
            displayName = initialProfile?.displayName ?? ""
            about = initialProfile?.about ?? ""
            nip05 = initialProfile?.nip05 ?? ""
            lud16 = initialProfile?.lud16 ?? ""
            picture = initialProfile?.picture ?? ""
            banner = initialProfile?.banner ?? ""
            website = initialProfile?.website ?? ""
        }
    }

    /// Legacy editor page chrome: back chevron + inline title.
    private var pageHeader: some View {
        HStack(spacing: BitOSTheme.Spacing.sm) {
            Button(action: onClose) {
                Image(systemName: "chevron.left")
                    .font(.system(size: 18, weight: .semibold))
                    .foregroundStyle(BitOSTheme.textPrimary)
                    .frame(width: 36, height: 36)
                    .contentShape(Rectangle())
            }
            .accessibilityLabel("Back")
            Text("Edit profile")
                .font(.system(size: 17, weight: .bold))
            Spacer()
        }
        .padding(.horizontal, BitOSTheme.Spacing.md)
        .padding(.vertical, BitOSTheme.Spacing.xs + 2)
    }

    /// Legacy `_chooseImageSource` parity: camera (when available) or
    /// photo library, as a bottom-sheet step before the picker.
    private func sourceSheet(target: String) -> some View {
        var entries: [AppMenuEntry] = [
            .item(AppMenuItem(id: "library", label: "Choose from library", systemImage: "photo.on.rectangle")),
        ]
        if UIImagePickerController.isSourceTypeAvailable(.camera) {
            entries.insert(.item(AppMenuItem(id: "camera", label: "Take photo", systemImage: AppIcons.camera)), at: 0)
        }
        return AppBottomSheetMenu(title: "Choose photo", entries: entries) { id in
            sourceSheetTarget = nil
            if id == "camera" {
                cameraTarget = target
                showCamera = true
            } else if target == "avatar" {
                showAvatarPicker = true
            } else {
                showBannerPicker = true
            }
        }
    }

    private var cameraPicker: some View {
        CameraPicker { data in
            showCamera = false
            if let data, let target = cameraTarget {
                cameraTarget = nil
                Task { await uploadPicked(target, data) }
            }
        }
    }

    @ViewBuilder
    private var content: some View {
        // Live header preview + legacy form card: single-column full-width
        // fields, compact system-style inputs, bio counter, full kind-0 set.
        ScrollView {
            VStack(spacing: BitOSTheme.Spacing.md) {
                headerPreview
                if published {
                    Label { Text("Profile published \u{2713}") } icon: { AppIcons.image(for: AppIcons.checkCircle) }
                        .foregroundStyle(BitOSTheme.success)
                        .frame(maxWidth: .infinity, alignment: .leading)
                } else if case .rejected(let detail) = publisher.result {
                    Text(detail ?? "Relays rejected the profile.").font(.caption).foregroundStyle(BitOSTheme.error)
                }

                field("Username", "username", $name)
                field("Display name", "Your name", $displayName)
                VStack(alignment: .leading, spacing: 4) {
                    fieldLabel("Bio")
                    FormTextField(placeholder: "Tell the world about yourself\u{2026}", text: $about, axis: .vertical)
                        .lineLimit(3)
                    Text("\(about.count) / 300 characters")
                        .font(.system(size: 11))
                        .foregroundStyle(BitOSTheme.textTertiary)
                }
                if let uploadError {
                    Text(uploadError).font(.system(size: 12)).foregroundStyle(BitOSTheme.error)
                }
                field(uploadingTarget == "avatar" ? "Uploading\u{2026} (avatar URL)" : "Avatar picture URL", "https://\u{2026}", $picture)
                field(uploadingTarget == "banner" ? "Uploading\u{2026} (banner URL)" : "Banner picture URL", "https://\u{2026}", $banner)
                field("Website", "https://example.com", $website)
                field("NIP-05", "name@example.com", $nip05)
                field("Lightning address", "name@getalby.com", $lud16)

                Text("Publishes a signed profile event to your relays; the change appears once confirmed.")
                    .font(.caption2)
                    .foregroundStyle(BitOSTheme.textTertiary)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.top, 4)

                // Legacy save: full-width pill, spinner while publishing.
                Button {
                    Task {
                        await publisher.publishProfile(
                            name: name, displayName: displayName, about: about,
                            picture: picture, nip05: nip05, lud16: lud16,
                            banner: banner, website: website
                        )
                        published = true
                    }
                } label: {
                    HStack(spacing: 8) {
                        if publisher.busy {
                            ProgressView()
                                .tint(.white)
                                .controlSize(.small)
                        }
                        Text(publisher.busy ? "Saving\u{2026}" : "Save changes")
                            .font(.system(size: 15, weight: .bold))
                    }
                    .frame(maxWidth: .infinity, minHeight: 48)
                }
                .buttonStyle(.plain)
                .foregroundStyle(.white)
                .background(Capsule(style: .continuous).fill(BitOSTheme.accent))
                // Media must be uploaded and hash-verified before signing.
                .disabled(publisher.busy || uploadingTarget != nil || name.isEmpty && displayName.isEmpty)
                .padding(.top, BitOSTheme.Spacing.sm)
            }
        }
    }

    // MARK: - Live header preview (legacy _ProfileHeaderPreview parity)

    /// Banner preview (120pt, radius 12) with the "Change banner" pill, the
    /// hex avatar (88pt, lifted -36) with its camera chip, and the identity
    /// row — exactly what the profile page will show once saved.
    private var headerPreview: some View {
        VStack(spacing: 0) {
            ZStack(alignment: .bottomTrailing) {
                bannerPreviewMedia
                    .frame(height: 120)
                    .frame(maxWidth: .infinity)
                    .clipShape(RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous))
                changeBannerPill
                    .padding(BitOSTheme.Spacing.sm)
            }
            VStack(spacing: BitOSTheme.Spacing.xs + 2) {
                avatarWithCameraChip
                identityPreviewRow
            }
            // Lift exactly half the avatar (88/2): the hexagon's center
            // sits ON the banner bottom edge — same line as the hero.
            .offset(y: -44)
            .padding(.bottom, -44)
        }
        .padding(.top, BitOSTheme.Spacing.sm)
        .padding(.bottom, BitOSTheme.Spacing.lg)
        .photosPicker(isPresented: $showAvatarPicker, selection: $avatarItem, matching: .images)
        .photosPicker(isPresented: $showBannerPicker, selection: $bannerItem, matching: .images)
        .onChange(of: avatarItem) { _, item in
            guard let item else { return }
            Task {
                if let data = try? await item.loadTransferable(type: Data.self) {
                    await uploadPicked("avatar", data)
                }
                avatarItem = nil
            }
        }
        .onChange(of: bannerItem) { _, item in
            guard let item else { return }
            Task {
                if let data = try? await item.loadTransferable(type: Data.self) {
                    await uploadPicked("banner", data)
                }
                bannerItem = nil
            }
        }
    }

    @ViewBuilder
    private var bannerPreviewMedia: some View {
        ZStack {
            if let url = URL(string: banner), !banner.isEmpty {
                AsyncImage(url: url) { image in
                    image.resizable().aspectRatio(contentMode: .fill)
                } placeholder: {
                    bannerPlaceholder
                }
            } else {
                bannerPlaceholder
            }
            if uploadingTarget == "banner" {
                Color.black.opacity(0.45)
                VStack(spacing: 6) {
                    ProgressView().tint(.white)
                    Text("Uploading")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(.white)
                }
            }
        }
    }

    private var bannerPlaceholder: some View {
        ZStack {
            RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous)
                .fill(BitOSTheme.accent.opacity(0.15))
            AppIcons.image(for: "photo.badge.plus")
                .font(.system(size: 40, weight: .medium))
                .foregroundStyle(BitOSTheme.accent.opacity(0.40))
        }
    }

    private var changeBannerPill: some View {
        Button {
            sourceSheetTarget = SourceTarget(id: "banner")
        } label: {
            HStack(spacing: 4) {
                AppIcons.image(for: AppIcons.camera)
                    .font(.system(size: 12, weight: .semibold))
                Text("Change banner")
                    .font(.system(size: 11, weight: .semibold))
            }
            .foregroundStyle(.white)
            .padding(.horizontal, 8)
            .padding(.vertical, 4)
            .background(RoundedRectangle(cornerRadius: BitOSTheme.Radius.sm).fill(.black.opacity(0.60)))
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Change banner photo")
    }

    private var avatarWithCameraChip: some View {
        Button {
            sourceSheetTarget = SourceTarget(id: "avatar")
        } label: {
            ZStack(alignment: .bottom) {
                HexAvatarView(
                    pubkey: identity.account?.pubkeyHex ?? "",
                    size: 88,
                    imageURL: safeProfilePictureURL(picture.isEmpty ? nil : picture),
                    label: displayName.isEmpty ? name : displayName
                )
                .shadow(color: .black.opacity(0.18), radius: 12, y: 5)
                .shadow(color: BitOSTheme.accent.opacity(0.14), radius: 8)
                // Camera chip — bottom-center-left, clear of the ⚡ slot.
                ZStack {
                    Circle().fill(BitOSTheme.accent)
                    Circle().strokeBorder(BitOSTheme.background, lineWidth: 2)
                    if uploadingTarget == "avatar" {
                        ProgressView()
                            .tint(.white)
                            .controlSize(.small)
                    } else {
                        AppIcons.image(for: AppIcons.camera)
                            .font(.system(size: 14, weight: .semibold))
                            .foregroundStyle(.white)
                    }
                }
                .frame(width: 28, height: 28)
                .offset(x: -14, y: 6)
            }
        }
        .buttonStyle(.plain)
        .accessibilityLabel(uploadingTarget == "avatar" ? "Uploading avatar" : "Change profile picture")
    }

    private var identityPreviewRow: some View {
        VStack(spacing: 2) {
            HStack(spacing: 4) {
                Text(displayName.isEmpty ? (name.isEmpty ? "Anonymous" : name) : displayName)
                    .font(.system(size: 20, weight: .heavy))
                    .lineLimit(1)
                if !nip05.isEmpty {
                    Image(systemName: AppIcons.checkCircle)
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(BitOSTheme.cyan)
                }
            }
            if !name.isEmpty {
                Text("@\(name)")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(BitOSTheme.accent)
            }
        }
    }

    private func fieldLabel(_ text: String) -> some View {
        Text(text.uppercased())
            .font(.system(size: 12, weight: .bold))
            .foregroundStyle(BitOSTheme.textSecondary)
    }

    private func field(_ label: String, _ placeholder: String, _ binding: Binding<String>) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            fieldLabel(label)
            FormTextField(placeholder: placeholder, text: binding)
        }
    }
}

private extension UIImage {
    /// Redraws non-up orientations into `.up` so pixel-space center-cropping
    /// (and the JPEG re-encode) matches what the user saw in the picker.
    func normalizedForCrop() -> UIImage {
        guard imageOrientation != .up else { return self }
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = scale
        return UIGraphicsImageRenderer(size: size, format: format).image { _ in
            draw(in: CGRect(origin: .zero, size: size))
        }
    }
}


/// Identifiable wrapper so `.sheet(item:)` can present the source sheet.
private struct SourceTarget: Identifiable {
    let id: String
}

/// System camera capture (legacy `ImageSource.camera` step). The shot
/// flows through the same crop → upload pipeline as library picks.
private struct CameraPicker: UIViewControllerRepresentable {
    let completion: (Data?) -> Void

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(completion: completion) }

    final class Coordinator: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        let completion: (Data?) -> Void

        init(completion: @escaping (Data?) -> Void) {
            self.completion = completion
        }

        func imagePickerController(
            _ picker: UIImagePickerController,
            didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]
        ) {
            let data = (info[.originalImage] as? UIImage)?.jpegData(compressionQuality: 0.92)
            completion(data)
        }

        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            completion(nil)
        }
    }
}


/// Compact brand input (legacy Flutter `InputDecorationTheme` parity):
/// filled surfaceElevated, 12° radius, 2pt accent focused border, 14pt
/// text, tight h12/v9 padding — the system-design field size.
struct FormTextField: View {
    let placeholder: String
    @Binding var text: String
    var axis: Axis = .horizontal
    @FocusState private var focused: Bool

    var body: some View {
        TextField(placeholder, text: $text, axis: axis == .vertical ? .vertical : .horizontal)
            .font(.system(size: 14))
            .foregroundStyle(BitOSTheme.textPrimary)
            .focused($focused)
            .padding(.horizontal, 12)
            .padding(.vertical, 9)
            .frame(minHeight: 44, alignment: axis == .vertical ? .top : .center)
            .background(
                RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous)
                    .fill(BitOSTheme.surfaceElevated)
            )
            .overlay(
                RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous)
                    .strokeBorder(focused ? BitOSTheme.accent : BitOSTheme.border, lineWidth: focused ? 2 : 1)
            )
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/ui/profile/ProfileEditSheet.kt
package space.bitos.app.ui.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import kotlinx.coroutines.launch
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.rounded.ArrowBack
import androidx.compose.material.icons.outlined.AddPhotoAlternate
import androidx.compose.material.icons.rounded.CheckCircle
import androidx.compose.material.icons.outlined.Upload
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import space.bitos.app.ui.components.AppBottomSheetMenu
import space.bitos.app.ui.components.AppMenuEntry
import space.bitos.app.ui.components.AppMenuItem
import space.bitos.app.ui.components.HexShape
import space.bitos.app.ui.components.PubkeyAvatar
import space.bitos.app.ui.theme.AppIcons
import space.bitos.app.ui.theme.BitOSColors
import space.bitos.app.ui.theme.BitOSSpacing

/**
 * Profile editing sheet: bounded fields → signed kind-0 through the receipt
 * machine. The relay echo (verified, newer) updates the profile projection.
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
fun ProfileEditContent(
    initialName: String = "",
    initialDisplayName: String = "",
    initialAbout: String = "",
    initialNip05: String = "",
    initialLud16: String = "",
    initialPicture: String = "",
    initialBanner: String = "",
    initialWebsite: String = "",
    pubkey: String = "",
    /** Page mode hides the sheet header (back lives in the page top bar). */
    showHeader: Boolean = true,
    error: String? = null,
    busy: Boolean,
    onUploadImage: (suspend (target: String, bytes: ByteArray) -> String?)? = null,
    onPublish: (name: String, displayName: String, about: String, nip05: String, lud16: String, picture: String, banner: String, website: String) -> Unit,
    onClose: () -> Unit,
) {
    var name by rememberSaveable { mutableStateOf(initialName) }
    var displayName by rememberSaveable { mutableStateOf(initialDisplayName) }
    var about by rememberSaveable { mutableStateOf(initialAbout) }
    var nip05 by rememberSaveable { mutableStateOf(initialNip05) }
    var lud16 by rememberSaveable { mutableStateOf(initialLud16) }
    var picture by rememberSaveable { mutableStateOf(initialPicture) }
    val context = androidx.compose.ui.platform.LocalContext.current
    var banner by rememberSaveable { mutableStateOf(initialBanner) }
    var website by rememberSaveable { mutableStateOf(initialWebsite) }
    val scope = androidx.compose.runtime.rememberCoroutineScope()
    var uploadingTarget by androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf<String?>(null) }
    var uploadError by androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf<String?>(null) }
    // Legacy step flow: pick target → source sheet (camera/library).
    var sourceTarget by androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf<String?>(null) }
    var cameraTarget by androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf<String?>(null) }

    suspend fun handlePicked(target: String, bytes: ByteArray?) {
        if (bytes == null || onUploadImage == null) return
        uploadingTarget = target
        uploadError = null
        val url = onUploadImage(target, bytes)
        uploadingTarget = null
        if (url == null) {
            uploadError = "Upload failed — check your connection and try again."
        } else {
            when (target) {
                "avatar" -> picture = url
                "banner" -> banner = url
            }
        }
    }

    // Photo pickers (legacy image_picker parity; center-crop inside upload).
    val avatarPicker = androidx.activity.compose.rememberLauncherForActivityResult(
        androidx.activity.result.contract.ActivityResultContracts.GetContent(),
    ) { uri ->
        val bytes = uri?.let { runCatching { context.contentResolver.openInputStream(it)?.use { stream -> stream.readBytes() } }.getOrNull() }
        scope.launch { handlePicked("avatar", bytes) }
    }
    val bannerPicker = androidx.activity.compose.rememberLauncherForActivityResult(
        androidx.activity.result.contract.ActivityResultContracts.GetContent(),
    ) { uri ->
        val bytes = uri?.let { runCatching { context.contentResolver.openInputStream(it)?.use { stream -> stream.readBytes() } }.getOrNull() }
        scope.launch { handlePicked("banner", bytes) }
    }

    val cameraLauncher = androidx.activity.compose.rememberLauncherForActivityResult(
        androidx.activity.result.contract.ActivityResultContracts.TakePicturePreview(),
    ) { bitmap ->
        val target = cameraTarget
        cameraTarget = null
        if (bitmap != null && target != null) {
            val stream = java.io.ByteArrayOutputStream()
            bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 92, stream)
            scope.launch { handlePicked(target, stream.toByteArray()) }
        }
    }

    @Composable
    fun FieldLabel(text: String) {
        Text(
            text.uppercase(),
            fontSize = 12.sp, fontWeight = FontWeight.W700,
            color = BitOSColors.textSecondary,
            modifier = Modifier.padding(bottom = 6.dp, top = 4.dp),
        )
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = BitOSSpacing.screen)
            .padding(bottom = BitOSSpacing.xl),
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        if (showHeader) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text("Edit profile", style = MaterialTheme.typography.headlineMedium)
                Spacer(Modifier.weight(1f))
                space.bitos.app.ui.components.SheetCloseIcon(onClose = onClose)
            }
        }

        // Legacy `_chooseImageSource` parity: camera or library step.
        sourceTarget?.let { target ->
            AppBottomSheetMenu(
                onDismissRequest = { sourceTarget = null },
                title = "Choose photo",
                entries = listOf(
                    AppMenuEntry.Item(AppMenuItem("camera", "Take photo", AppIcons.Camera)),
                    AppMenuEntry.Item(AppMenuItem("library", "Choose from library", AppIcons.Photo)),
                ),
                onSelect = { id ->
                    sourceTarget = null
                    if (id == "camera") {
                        cameraTarget = target
                        cameraLauncher.launch(null)
                    } else if (target == "avatar") {
                        avatarPicker.launch("image/*")
                    } else {
                        bannerPicker.launch("image/*")
                    }
                },
            )
        }

        error?.let { err ->
            Surface(shape = RoundedCornerShape(12.dp), color = BitOSColors.surface) {
                Text(
                    err,
                    style = MaterialTheme.typography.bodySmall,
                    color = BitOSColors.error,
                    modifier = Modifier.padding(BitOSSpacing.base),
                )
            }
        }

        // ── Live header preview (legacy _ProfileHeaderPreview parity) ──
        EditHeaderPreview(
            pubkey = pubkey,
            name = name,
            displayName = displayName,
            nip05 = nip05,
            picture = picture,
            banner = banner,
            uploadingTarget = uploadingTarget,
            onPickAvatar = { sourceTarget = "avatar" },
            onPickBanner = { sourceTarget = "banner" },
        )

        // ── Legacy form card: single-column full-width compact fields. ──
        Column(Modifier.fillMaxWidth().padding(top = 8.dp)) {
            FieldLabel("Username")
            space.bitos.app.ui.components.BitosTextField(
                value = name,
                onValueChange = { if (it.length <= 64) name = it },
                placeholder = "username",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel("Display name")
            space.bitos.app.ui.components.BitosTextField(
                value = displayName,
                onValueChange = { if (it.length <= 64) displayName = it },
                placeholder = "Your name",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel("Bio")
            space.bitos.app.ui.components.BitosTextField(
                value = about,
                onValueChange = { if (it.length <= 300) about = it },
                placeholder = "Tell the world about yourself…",
                singleLine = false,
                minLines = 3,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
            Text("${'$'}{about.length} / 300 characters", fontSize = 11.sp, color = BitOSColors.textTertiary, modifier = Modifier.padding(top = 4.dp))
        }
        uploadError?.let { Text(it, fontSize = 12.sp, color = BitOSColors.error) }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel(if (uploadingTarget == "avatar") "Uploading… (avatar URL)" else "Avatar picture URL")
            space.bitos.app.ui.components.BitosTextField(
                value = picture,
                onValueChange = { if (it.length <= 256) picture = it },
                placeholder = "https://…",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel(if (uploadingTarget == "banner") "Uploading… (banner URL)" else "Banner picture URL")
            space.bitos.app.ui.components.BitosTextField(
                value = banner,
                onValueChange = { if (it.length <= 256) banner = it },
                placeholder = "https://…",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel("Website")
            space.bitos.app.ui.components.BitosTextField(
                value = website,
                onValueChange = { if (it.length <= 128) website = it },
                placeholder = "https://example.com",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel("NIP-05")
            space.bitos.app.ui.components.BitosTextField(
                value = nip05,
                onValueChange = { if (it.length <= 64) nip05 = it },
                placeholder = "name@example.com",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Column(Modifier.fillMaxWidth()) {
            FieldLabel("Lightning address")
            space.bitos.app.ui.components.BitosTextField(
                value = lud16,
                onValueChange = { if (it.length <= 64) lud16 = it },
                placeholder = "name@getalby.com",
                singleLine = true,
                compact = true,
                modifier = Modifier.fillMaxWidth(),
            )
        }

        Text(
            "Publishes a signed profile event to your relays; the change appears once confirmed.",
            style = MaterialTheme.typography.labelSmall,
            color = BitOSColors.textTertiary,
            modifier = Modifier.padding(top = 8.dp),
        )

        // Legacy save: full-width pill with spinner; the page back cancels.
        Button(
            onClick = { onPublish(name, displayName, about, nip05, lud16, picture, banner, website) },
            enabled = !busy && uploadingTarget == null,
            shape = RoundedCornerShape(16.dp),
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 12.dp)
                .height(48.dp),
        ) {
            if (busy) {
                androidx.compose.material3.CircularProgressIndicator(
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(20.dp),
                )
                Spacer(Modifier.width(10.dp))
            }
            Text(if (busy) "Saving…" else "Save changes", fontSize = 15.sp, fontWeight = FontWeight.W700)
        }
    }
}

/// Live preview of the profile hero exactly as it will render once saved
/// (legacy Flutter `_ProfileHeaderPreview` parity): banner 120 with the
/// "Change banner" pill, hex avatar 88 lifted -36 with a camera chip, and
/// the identity row (name + verified + @handle).
@Composable
private fun EditHeaderPreview(
    pubkey: String,
    name: String,
    displayName: String,
    nip05: String,
    picture: String,
    banner: String,
    uploadingTarget: String?,
    onPickAvatar: () -> Unit,
    onPickBanner: () -> Unit,
) {
    Box(Modifier.fillMaxWidth()) {
        // Banner preview (or brand placeholder).
        Box(
            Modifier
                .align(Alignment.TopCenter)
                .fillMaxWidth()
                .padding(top = 4.dp)
                .height(120.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(BitOSColors.primary.copy(alpha = 0.15f)),
        ) {
            if (banner.isNotBlank()) {
                coil.compose.AsyncImage(
                    model = banner,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.fillMaxSize(),
                )
            } else {
                Icon(
                    Icons.Outlined.AddPhotoAlternate,
                    contentDescription = null,
                    tint = BitOSColors.primary.copy(alpha = 0.40f),
                    modifier = Modifier.size(40.dp).align(Alignment.Center),
                )
            }
            if (uploadingTarget == "banner") {
                Box(Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.45f)))
                Column(
                    Modifier.align(Alignment.Center),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(26.dp))
                    Text("Uploading", fontSize = 11.sp, fontWeight = FontWeight.W600, color = Color.White)
                }
            }
            // "Change banner" pill — bottom-right, black/60 glass.
            Row(
                Modifier
                    .align(Alignment.BottomEnd)
                    .padding(8.dp)
                    .clip(RoundedCornerShape(8.dp))
                    .background(Color.Black.copy(alpha = 0.60f))
                    .clickable(onClickLabel = "Change banner photo") { onPickBanner() }
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                Icon(AppIcons.Camera, contentDescription = null, tint = Color.White, modifier = Modifier.size(12.dp))
                Text("Change banner", fontSize = 11.sp, fontWeight = FontWeight.W600, color = Color.White)
            }
        }
        // Avatar + camera chip: lift exactly half the avatar (88/2) so the
        // hexagon's center sits ON the banner bottom edge (banner top pad 4
        // + 120 = 124; avatar top = 124 - 44 = 80) — the hero center line.
        Column(
            Modifier
                .align(Alignment.TopCenter)
                .offset(y = (124 - 44).dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Box(
                Modifier
                    .size(88.dp)
                    .shadow(8.dp, HexShape())
                    .clickable(onClickLabel = "Change profile picture") { onPickAvatar() },
                contentAlignment = Alignment.BottomCenter,
            ) {
                PubkeyAvatar(
                    pubkey = pubkey,
                    size = 88,
                    pictureUrl = picture.ifBlank { null },
                    label = displayName.ifBlank { name },
                )
                // Camera chip — bottom-left of center, clear of the ⚡ slot.
                Box(
                    Modifier
                        .offset(x = (-14).dp, y = 6.dp)
                        .size(28.dp)
                        .clip(CircleShape)
                        .background(BitOSColors.primary)
                        .border(2.dp, BitOSColors.background, CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    if (uploadingTarget == "avatar") {
                        CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(14.dp), color = Color.White)
                    } else {
                        Icon(AppIcons.Camera, contentDescription = "Change profile picture", tint = Color.White, modifier = Modifier.size(14.dp))
                    }
                }
            }
            // Identity preview row.
            Column(
                Modifier.padding(top = 6.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        displayName.ifBlank { name.ifBlank { "Anonymous" } },
                        fontSize = 20.sp, fontWeight = FontWeight.W800, color = BitOSColors.textPrimary,
                    )
                    if (nip05.isNotBlank()) {
                        androidx.compose.foundation.layout.Spacer(Modifier.size(4.dp))
                        Icon(
                            Icons.Rounded.CheckCircle,
                            contentDescription = "Verified",
                            tint = BitOSColors.accent,
                            modifier = Modifier.size(14.dp),
                        )
                    }
                }
                if (name.isNotBlank()) {
                    Text("@$name", fontSize = 12.sp, fontWeight = FontWeight.W600, color = BitOSColors.primary)
                }
            }
        }
    }
}
